File: Python/LeetCodeTest/lc084_LargestRectangleInHistogram.py
from typing import List

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        if not heights:
            return 0
        m = len(heights)
        less_from_left = [0] * m
        less_from_left[0] = -1
        i = 1
        while i < m:
            p = i - 1
            while p >= 0 and heights[p] >= heights[i]:
                p = less_from_left[p]
            less_from_left[i] = p
            i += 1
        less_from_right = [0] * m
        less_from_right[m - 1] = m
        i = m - 2
        while i >= 0:
            p = i + 1
            while p < m and heights[p] >= heights[i]:
                p = less_from_right[p]
            less_from_right[i] = p
            i -= 1
        result = 0
        for i in range(0, m):
            result = max(result, heights[i] * (less_from_right[i] - less_from_left[i] - 1))
        return result

if __name__ == "__main__":
    solution = Solution()
    result = solution.largestRectangleArea([2,1,5,6,2,3])
    print(result)

File: Java/src/com/liu/LeetCode/FifteenPatterns/MonotonicStack/LC084_LargestRectangleInHistogram.java
package com.liu.LeetCode.FifteenPatterns.MonotonicStack;

import java.util.Stack;

/**
 * The Monotonic Stack pattern uses a stack to maintain a sequence of elements in a specific order (increasing or decreasing).
 * Use this pattern for problems that require finding the next greater or smaller element.
 * Sample Problem:
 * Find the next greater element for each element in an array. Output -1 if the greater element doesn’t exist.
 * Example:
 * Input: nums = [2, 1, 2, 4, 3]
 * Output: [4, 2, 4, -1, -1]
 * Explanation:
 * Use a stack to keep track of elements for which we haven't found the next greater element yet.
 * Iterate through the array, and for each element, pop elements from the stack until you find a greater element.
 * If the stack is not empty, set the result for index at the top of the stack to current element.
 * Push the current element onto the stack.
 */

public class LC084_LargestRectangleInHistogram {
    private static int largestRectangleAreaMy1(int[] heights) { // accepted 性能不好
        int result = 0;
        int length = heights.length;
        Stack<Integer> stack = new Stack<>();
        stack.push(-1); // 处理左侧
        for (int i = 0; i < length + 1; i++) {
            int height;
            if (i == length) {
                height = 0; // 处理右侧
            } else {
                height = heights[i];
            }
            while (stack.peek() != -1 && heights[stack.peek()] > height) {
                int leftHeight = heights[stack.pop()];
                int leftWidth = i - stack.peek() - 1;
                result = Math.max(result, leftHeight * leftWidth);
            }
            stack.push(i);
        }
        return result;
    }

    public static void main(String[] args) {
        int[][] numsSet = new int[][]{
                {2, 1, 5, 6, 2, 3},
                {2, 4, 6, 8},
                {8, 6, 4, 2},
        };
        int result;
        long startTime;
        long endTime;

        for (int i = 0; i < numsSet.length; i++) {
            startTime = System.currentTimeMillis();
            result = largestRectangleAreaMy1(numsSet[i]);
            endTime = System.currentTimeMillis();
            System.out.println("largestRectangleArea My1 result" + i + " " + result + " during time " + (endTime - startTime));
        }
    }
}
